emma-language/src/main/scala/org/emmalanguage/compiler/lang/opt/FoldGroupFusion.scala

```diff
@@ -0,0 +1,409 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.emmalanguage
+package compiler.lang.opt
+
+import compiler.Common
+import compiler.ir.Aggregations
+import compiler.lang.cf.ControlFlow
+import compiler.lang.core.Core
+import util.Functions
+import util.Monoids._
+import util.Graphs._
+
+import cats.std.all._
+import scalaz.Tree
+import shapeless._
+
+import scala.collection.breakOut
+
+/** The fold-group-fusion optimization. */
+private[compiler] trait FoldGroupFusion extends Common {
+  this: Core with ControlFlow =>
+
+  /** The fold-group-fusion optimization. */
+  object FoldGroupFusion {
+
+    import Aggregations._
+    import UniverseImplicits._
+    import Core.{Lang => core}
+
+    // Function utils.
+    private val functions = core.Ref(api.Sym[Functions.type].asModule)
+    private val conditional = functions.tpe.member(api.TermName("conditional")).asMethod
+    private val refold = functions.tpe.member(api.TermName("refold")).asMethod
+
+    // Folds without arguments.
+    private val noArgFolds = Map(
+      API.isEmpty -> api.Sym[IsEmpty.type],
+      API.nonEmpty -> api.Sym[NonEmpty.type],
+      API.size -> api.Sym[Size.type]
+    ).mapValues(_.asModule)
+
+    // Folds that require no tally of the result.
+    private val noTallyFolds = Map(
+      API.count -> api.Sym[Count.type],
+      API.sum -> api.Sym[Sum.type],
+      API.product -> api.Sym[Product.type],
+      API.exists -> api.Sym[Exists.type],
+      API.forall -> api.Sym[Forall.type],
+      API.find -> api.Sym[Find.type],
+      API.bottom -> api.Sym[Bottom.type],
+      API.top -> api.Sym[Top.type],
+      API.reduceOption -> api.Sym[ReduceOpt.type]
+    ).mapValues(_.asModule)
+
+    // Folds with an optional result that has to be extracted.
+    private val optionFolds = Map(
+      API.min -> api.Sym[Min.type],
+      API.max -> api.Sym[Aggregations.Max.type]
+    ).mapValues(_.asModule)
+
+    /**
+     * Given a sequence of vals in a block and the global control-flow graph,
+     * returns a sequence of DAGs that can be fused in a single fold.
+     */
+    private def foldForest(vals: Map[u.TermSymbol, u.Tree], cfg: CFG.FlowGraph[u.TermSymbol])
+      : Seq[Tree[u.TermSymbol]] = {
+        val CFG.FlowGraph(refCount, _, _, dataFlow) = cfg
+
+        // Initial fold forest.
+        val forest: Stream[Tree[u.TermSymbol]] = vals.collect {
+          case (fold, core.DefCall(_, method, _, _))
+            if API.foldOps(method) => Tree.leaf(fold)
+        } (breakOut)
+
+        // Iterate to combine trees in the forest while possible.
+        // We can't forget about trees that don't change in a single step,
+        // because a squash might enable a fusion and vice versa.
+        Iterator.iterate(forest, true) { case (forest, _) =>
```

Comment:
OK, the `flatMap` is on a Stream because of the preceeding `toStream` call, maybe just do a `flatMap` on the result of the `groupBy` (a `Map`) and convert to `Stream` at the end.